File: AOC2017/Day1/FSharp/AOC2017-Day1/Program.fs
open System
open System.IO

let introPath = @"D:\AdventOfCode\AOC2017\Day1\Inputs\intro.txt"

let actualPath = @"D:\AdventOfCode\AOC2017\Day1\Inputs\actual.txt"

let ParseInt(c:char) =
    Double.Parse(c.ToString()) |> int

let Intro() =
    let text = File.ReadAllText(introPath)
    let len = text.Length - 1
    let loop = text + text
    let mutable sum = 0
    for i = 0 to len do
        if ParseInt(loop[i]) = ParseInt(loop[i+1]) then
            sum <- sum + ParseInt(loop[i])
    sum
    
    
let PartOne() =
    let text = File.ReadAllText(actualPath)
    let len = text.Length - 1
    let loop = text + text
    let mutable sum = 0
    for i = 0 to len do
        if ParseInt(loop[i]) = ParseInt(loop[i+1]) then
            sum <- sum + ParseInt(loop[i])
    sum
    
let PartTwo() =
    let text = File.ReadAllText(actualPath)
    let len = text.Length - 1
    let half = text.Length / 2
    let loop = text + text
    let mutable sum = 0
    for i = 0 to len do
        if ParseInt(loop[i]) = ParseInt(loop[i + half]) then
            sum <- sum + ParseInt(loop[i])
    sum
    
[<EntryPoint>]
let Main(args) =   
    Intro() |> printfn "Intro: %d" 
    PartOne() |> printfn "Part One: %d" 
    PartTwo() |> printfn "Part Two: %d" 
    0

File: AOC2017/Day1/CSharp/AOC2017-Day1/Program.cs
namespace AOC2017_Day1; 

static class Program {
    
    private const string Intro = @"D:\AdventOfCode\AOC2017\Day1\Inputs\intro.txt";
    private const string Actual = @"D:\AdventOfCode\AOC2017\Day1\Inputs\actual.txt";
    
    public static void Main() {
        Console.WriteLine($"Intro Sum: {PartIntro()}");
        Console.WriteLine($"Part One Sum: {PartOne()}");
        Console.WriteLine($"Part Two Sum: {PartTwo()}");
    }

    private static int PartIntro() {
        var text = File.ReadAllText(Intro);
        var sum = 0;
        for (var i = 0; i < text.Length-1; i++) {
            var cur = (int)char.GetNumericValue(text[i]);
            var next = (int)char.GetNumericValue(text[i+1]);
            if (cur == next) {
                sum += cur;
            }
        }

        if ((int)char.GetNumericValue(text.First()) == (int)char.GetNumericValue(text.Last())) {
            sum += (int)char.GetNumericValue(text.First());
        }

        return sum;
    }
    private static int PartOne() {
        var text = File.ReadAllText(Actual);
        var sum = 0;
        for (var i = 0; i < text.Length-1; i++) {
            var cur = (int)char.GetNumericValue(text[i]);
            var next = (int)char.GetNumericValue(text[i+1]);
            if (cur == next) {
                sum += cur;
            }
        }

        if ((int)char.GetNumericValue(text.First()) == (int)char.GetNumericValue(text.Last())) {
            sum += (int)char.GetNumericValue(text.First());
        }

        return sum;
    }
    
    private static int PartTwo() {
        var text = File.ReadAllText(Actual);
        var loop = text + text;
        var sum = 0;
        var half = text.Length / 2;
        for (var i = 0; i < text.Length; i++) {
            var cur = (int)char.GetNumericValue(loop[i]);
            var next = (int)char.GetNumericValue(loop[i+half]);
            if (cur == next) {
                sum += cur;
            }
        }
        return sum;
    }
    
}
